File: src/android/app/src/main/java/com/mintreplica/lite/data/database/entities/GoalEntity.kt
/**
 * Human Tasks:
 * 1. Verify Room database schema version is properly configured in AppDatabase
 * 2. Ensure DateConverter is registered in the AppDatabase configuration
 */

package com.mintreplica.lite.data.database.entities

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import androidx.room.TypeConverters
import com.mintreplica.lite.domain.model.Goal
import java.math.BigDecimal
import java.util.Date

/**
 * Room database entity representing a financial goal in the local SQLite database.
 * 
 * Requirements addressed:
 * - Financial Goal Setting and Progress Monitoring (1.2 Scope/Core Features)
 * - Local Data Storage (5.2.1 Mobile Applications)
 * - Goal Data Model (8.2.1 Schema Design/Goals)
 */
@Entity(tableName = "goals")
@TypeConverters(DateConverter::class)
data class GoalEntity(
    @PrimaryKey
    @ColumnInfo(name = "id")
    val id: String,

    @ColumnInfo(name = "user_id")
    val userId: String,

    @ColumnInfo(name = "name")
    val name: String,

    @ColumnInfo(name = "description")
    val description: String,

    @ColumnInfo(name = "target_amount")
    val targetAmount: BigDecimal,

    @ColumnInfo(name = "current_amount")
    val currentAmount: BigDecimal,

    @ColumnInfo(name = "category")
    val category: String,

    @ColumnInfo(name = "target_date")
    val targetDate: Date,

    @ColumnInfo(name = "created_at")
    val createdAt: Date,

    @ColumnInfo(name = "updated_at")
    val updatedAt: Date,

    @ColumnInfo(name = "is_completed")
    val isCompleted: Boolean
) {
    /**
     * Converts the database entity to a domain model instance.
     * Implementation addresses requirement from 5.2.1 Mobile Applications for data mapping.
     *
     * @return Goal domain model instance
     */
    fun toDomainModel(): Goal {
        return Goal(
            id = id,
            userId = userId,
            name = name,
            description = description,
            targetAmount = targetAmount,
            currentAmount = currentAmount,
            category = category,
            targetDate = targetDate,
            createdAt = createdAt,
            updatedAt = updatedAt,
            isCompleted = isCompleted
        )
    }

    companion object {
        /**
         * Creates a database entity from a domain model instance.
         * Implementation addresses requirement from 5.2.1 Mobile Applications for data persistence.
         *
         * @param goal Domain model instance to convert
         * @return GoalEntity database entity instance
         */
        fun fromDomainModel(goal: Goal): GoalEntity {
            return GoalEntity(
                id = goal.id,
                userId = goal.userId,
                name = goal.name,
                description = goal.description,
                targetAmount = goal.targetAmount,
                currentAmount = goal.currentAmount,
                category = goal.category,
                targetDate = goal.targetDate,
                createdAt = goal.createdAt,
                updatedAt = goal.updatedAt,
                isCompleted = goal.isCompleted
            )
        }
    }
}

File: src/ios/MintReplicaLite/Data/Persistence/CoreData/Entities/GoalEntity.swift
//
// GoalEntity.swift
// MintReplicaLite
//
// Core Data entity for financial goal persistence
//

import CoreData // iOS 14.0+
import Foundation // iOS 14.0+

// MARK: - Human Tasks
/*
 1. Verify Core Data model schema matches entity attributes
 2. Test data migration scenarios for schema updates
 3. Configure Core Data model versioning
 4. Review index configuration for performance optimization
 5. Test concurrent Core Data operations
*/

/// Core Data entity class representing a financial goal in local storage
/// Addresses requirements:
/// - Financial goal setting and progress monitoring
/// - Local SQLite database for offline data
@objc(GoalEntity)
@objcMembers
public class GoalEntity: NSManagedObject {
    
    // MARK: - Properties
    
    /// Unique identifier of the goal
    @NSManaged public var id: UUID
    
    /// Display name of the goal
    @NSManaged public var name: String
    
    /// Detailed description of the goal
    @NSManaged public var goalDescription: String
    
    /// Target amount to achieve
    @NSManaged public var targetAmount: NSDecimalNumber
    
    /// Current progress amount
    @NSManaged public var currentAmount: NSDecimalNumber
    
    /// Target completion date
    @NSManaged public var deadline: Date
    
    /// Creation timestamp
    @NSManaged public var createdAt: Date
    
    /// Completion timestamp (optional)
    @NSManaged public var completedAt: Date?
    
    /// Current goal status
    @NSManaged public var status: String
    
    /// Goal category
    @NSManaged public var category: String
    
    /// Last update timestamp
    @NSManaged public var updatedAt: Date
    
    // MARK: - Initialization
    
    /// Initializes a new goal entity with default values
    /// - Parameter context: NSManagedObjectContext for Core Data operations
    public override init(entity: NSEntityDescription, insertInto context: NSManagedObjectContext?) {
        super.init(entity: entity, insertInto: context)
        
        // Set default values
        self.id = UUID()
        self.createdAt = Date()
        self.updatedAt = Date()
        self.status = GoalStatus.notStarted.rawValue
        self.targetAmount = NSDecimalNumber.zero
        self.currentAmount = NSDecimalNumber.zero
    }
    
    // MARK: - Domain Model Conversion
    
    /// Converts Core Data entity to domain model
    /// - Returns: Goal domain model instance
    public func toDomain() -> Goal {
        // Convert stored amounts to Decimal
        let target = Decimal(string: targetAmount.stringValue) ?? .zero
        let current = Decimal(string: currentAmount.stringValue) ?? .zero
        
        // Create domain model instance with stored values
        do {
            let goal = try Goal(
                id: id,
                name: name,
                description: goalDescription,
                targetAmount: target,
                currentAmount: current,
                deadline: deadline,
                category: GoalCategory(rawValue: category) ?? .savings
            )
            return goal
        } catch {
            // Return default goal if validation fails
            // This should not happen with properly persisted data
            do {
                return try Goal(
                    id: id,
                    name: name,
                    description: goalDescription,
                    targetAmount: 1,
                    currentAmount: 0,
                    deadline: Date().addingTimeInterval(86400 * 30),
                    category: .savings
                )
            } catch {
                fatalError("Failed to create valid goal model: \(error)")
            }
        }
    }
    
    /// Updates entity with domain model data
    /// - Parameter goal: Goal domain model instance
    public func update(with goal: Goal) {
        // Update basic properties
        self.name = goal.name
        self.goalDescription = goal.description
        self.deadline = goal.deadline
        self.category = goal.category.rawValue
        self.status = goal.status.rawValue
        self.completedAt = goal.completedAt
        
        // Convert and update amounts
        self.targetAmount = NSDecimalNumber(decimal: goal.targetAmount)
        self.currentAmount = NSDecimalNumber(decimal: goal.currentAmount)
        
        // Update timestamp
        self.updatedAt = Date()
        
        // Mark object as updated in context
        if let context = self.managedObjectContext {
            context.refresh(self, mergeChanges: true)
        }
    }
}